velox/exec/LocalPartition.cpp

```diff
@@ -438,49 +438,137 @@ void LocalPartition::copy(
   target->copyRanges(input.get(), ranges);
 }
 
-RowVectorPtr LocalPartition::processPartition(
+void LocalPartition::populatePartitionBuffer(
     const RowVectorPtr& input,
-    vector_size_t size,
+    vector_size_t numPartitionRows,
     int partition,
-    const BufferPtr& indices,
-    const vector_size_t* rawIndices) {
+    const vector_size_t* rawIndices,
+    uint64_t& totalPartitionBufferSizeExcludingString,
+    uint64_t& totalPartitionStringBufferSize) {
+  VELOX_CHECK_GT(singlePartitionBufferSize_, 0);
+  if (partitionBuffers_.empty()) {
+    partitionBuffers_.resize(numPartitions_);
+  }
+  if (copyRanges_.size() < numPartitionRows) {
+    copyRanges_.resize(numPartitionRows);
+  }
+
+  auto& partitionBuffer = partitionBuffers_[partition];
+  auto targetIndex = 0;
+  if (partitionBuffer) {
+    targetIndex = partitionBuffer->size();
+  }
+  for (int i = 0; i < numPartitionRows; i++) {
+    copyRanges_[i] = {rawIndices[i], targetIndex, 1};
+    targetIndex++;
+  }
+
+  copy(
+      input,
+      folly::Range{copyRanges_.data(), static_cast<size_t>(numPartitionRows)},
```

Comment:
nit: Just do `copy(input, copyRanges_, partitionBuffer)` should be enough once you resize the `copyRanges_` unconditionally